internal/cobraprompt/prompt.go

```diff
@@ -0,0 +1,164 @@
+	// the value will be provided to the DynamicSuggestionsFunc function.
+	DynamicSuggestionsFunc func(annotationValue string, document *prompt.Document) []prompt.Suggest
+	// PersistFlagValues will persist flags. For example have verbose turned on every command.
+	PersistFlagValues bool
+	// FlagsToExclude is a list of flag names to specify flags to exclude from suggestions
+	FlagsToExclude []string
+	// Suggests flags without user type "-"
+	SuggestFlagsWithoutDash bool
+	// ShowHelpCommandAndFlags will make help command and flag for every command available.
+	ShowHelpCommandAndFlags bool
+	// DisableCompletionCommand will disable the default completion command for cobra
+	DisableCompletionCommand bool
+	// ShowHiddenCommands makes hidden commands available
+	ShowHiddenCommands bool
+	// ShowHiddenFlags makes hidden flags available
+	ShowHiddenFlags bool
+	// AddDefaultExitCommand adds a command for exiting prompt loop
+	AddDefaultExitCommand bool
+	// OnErrorFunc handle error for command.Execute, if not set print error and exit
+	OnErrorFunc func(err error)
+}
+
+// Run will automatically generate suggestions for all cobra commands and flags defined by RootCmd
+// and execute the selected commands. Run will also reset all given flags by default, see PersistFlagValues
+func (co CobraPrompt) Run(ctx context.Context) {
+	if co.RootCmd == nil {
+		panic("RootCmd is not set. Please set RootCmd")
+	}
+	co.prepare()
+	p := prompt.New(
+		func(in string) {
+			// do not execute root command if no input given
+			if in == "" {
+				return
+			}
+			promptArgs, err := shlex.Split(in)
+			if err != nil {
+				fmt.Println("unable to parse commands")
+				return
+			}
+			os.Args = append([]string{os.Args[0]}, promptArgs...)
+			if err := co.RootCmd.ExecuteContext(ctx); err != nil {
+				if co.OnErrorFunc != nil {
+					co.OnErrorFunc(err)
+				} else {
+					co.RootCmd.PrintErrln(err)
+					os.Exit(1)
+				}
+			}
+		},
+		func(d prompt.Document) []prompt.Suggest {
+			return findSuggestions(&co, &d)
+		},
+		co.GoPromptOptions...,
+	)
+	p.Run()
+}
+
+func (co CobraPrompt) prepare() {
+	if co.ShowHelpCommandAndFlags {
+		co.RootCmd.InitDefaultHelpCmd()
+	}
+	if co.DisableCompletionCommand {
+		co.RootCmd.CompletionOptions.DisableDefaultCmd = true
+	}
+	if co.AddDefaultExitCommand {
+		co.RootCmd.AddCommand(&cobra.Command{
+			Use:   "exit",
+			Short: "Exit prompt",
+			Run: func(cmd *cobra.Command, args []string) {
+				os.Exit(0)
+			},
+		})
+	}
+	if co.PersistFlagValues {
+		co.RootCmd.PersistentFlags().BoolP(PersistFlagValuesFlag, "",
+			false, "Persist last given value for flags")
+	}
+}
+
+func findSuggestions(co *CobraPrompt, d *prompt.Document) []prompt.Suggest {
+	command := co.RootCmd
+	args := strings.Fields(d.CurrentLine())
+	if found, _, err := command.Find(args); err == nil {
+		command = found
+	}
+	var suggestions []prompt.Suggest
+	persistFlagValues, _ := command.Flags().GetBool(PersistFlagValuesFlag)
```

Comment:
The error is ignored here. Is that intended?

Reply:
I tried to add a workaround, because initiation of commands etc.. are a mess because of the global state. Changes are on [28bd268](https://github.com/hazelcast/hazelcast-commandline-client/pull/32/commits/28bd268ea81423605d78b702bc754105544cbd3c). I will keep this in mind for a refactor session if that is okay with you too.